lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,1115 @@
+use bitcoin::network::constants::Network;
+use bitcoin::secp256k1::{Message, PublicKey, self};
+use bitcoin::secp256k1::schnorr::Signature;
+use core::convert::TryFrom;
+use crate::io;
+use crate::ln::features::InvoiceRequestFeatures;
+use crate::ln::msgs::DecodeError;
+use crate::offers::merkle::{SignatureTlvStream, SignatureTlvStreamRef, self};
+use crate::offers::offer::{Amount, Offer, OfferContents, OfferTlvStream, OfferTlvStreamRef};
+use crate::offers::parse::{ParseError, ParsedMessage, SemanticError};
+use crate::offers::payer::{PayerContents, PayerTlvStream, PayerTlvStreamRef};
+use crate::util::ser::{HighZeroBytesDroppedBigSize, SeekReadable, WithoutLength, Writeable, Writer};
+use crate::util::string::PrintableString;
+
+use crate::prelude::*;
+
+const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice_request", "signature");
+
+/// Builds an [`InvoiceRequest`] from an [`Offer`] for the "offer to be paid" flow.
+///
+/// See [module-level documentation] for usage.
+///
+/// [module-level documentation]: self
+pub struct InvoiceRequestBuilder<'a> {
+	offer: &'a Offer,
+	invoice_request: InvoiceRequestContents,
+}
+
+impl<'a> InvoiceRequestBuilder<'a> {
+	pub(super) fn new(offer: &'a Offer, metadata: Vec<u8>, payer_id: PublicKey) -> Self {
+		Self {
+			offer,
+			invoice_request: InvoiceRequestContents {
+				payer: PayerContents(metadata), offer: offer.contents.clone(), chain: None,
+				amount_msats: None, features: InvoiceRequestFeatures::empty(), quantity: None,
+				payer_id, payer_note: None,
+			},
+		}
+	}
+
+	/// Sets the [`InvoiceRequest::chain`] of the given [`Network`] for paying an invoice. If not
+	/// called, [`Network::Bitcoin`] is assumed. Must be supported by the offer.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn chain(mut self, network: Network) -> Self {
+		self.invoice_request.chain = Some(ChainHash::using_genesis_block(network));
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::amount_msats`] for paying an invoice. Must be at least the base
+	/// invoice amount (i.e., [`Offer::amount`] times [`quantity`]).
+	///
+	/// Successive calls to this method will override the previous setting.
+	///
+	/// [`quantity`]: Self::quantity
+	pub fn amount_msats(mut self, amount_msats: u64) -> Self {
+		self.invoice_request.amount_msats = Some(amount_msats);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::features`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	#[cfg(test)]
+	pub fn features(mut self, features: InvoiceRequestFeatures) -> Self {
+		self.invoice_request.features = features;
+		self
+	}
+
+	/// Sets [`InvoiceRequest::quantity`] of items. If not set, `1` is assumed. Must conform to
+	/// [`Offer::is_valid_quantity`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn quantity(mut self, quantity: u64) -> Result<Self, SemanticError> {
+		self.invoice_request.offer.check_quantity(Some(quantity))?;
```

Comment:
Ok, checking `amount_msats` when setting is problematic because we don't check the `Offer` currency until calling `build`. We can duplicated the check, of course. But thinking about this more, it seems strange to emit `UnsupportedCurrency` and `InsufficientAmount` when conditionally performing the same checks when setting `quantity`.

For consistency, I'm starting to prefer my original approach of only checking in `build`. Alternatively, we could be somewhat inconsistent and only check in setters when:
- all information is available
- but only for errors related to the field being set

(e.g., `quantity`, `chain`, `amount_msats` when `quantity` is already set or not expected)

But otherwise not.

(e.g., `amount_msats` when `quantity` is expected but hasn't been set, `amount_msats` when setting `quantity`)

What do ya'll think?

Reply:
Checking it in a `new` would work, too. Though it's nice to have the currency checked in the same place given the other checks involve calling the code with the unreachable path, which could be eliminated if combined. Pushed an untested version for consideration. May need to add more test-only unchecked setters.

Reply:
Ah, sorry. Meant that I pushed a version that does the checks in the `amount_msats` and `build` such that they are reused in both places and in the parsing code. This removes the `unreachable` code from `Offer`.

Reply:
This also now fully tested.